src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs

```diff
@@ -37,42 +36,34 @@ namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions
     /// </summary>
     internal struct RegexLexer
     {
-        public readonly ImmutableArray<VirtualChar> Text;
+        public readonly VirtualCharSequence Text;
         public int Position;
 
-        public RegexLexer(ImmutableArray<VirtualChar> text) : this()
+        public RegexLexer(VirtualCharSequence text) : this()
         {
             Text = text;
         }
 
         public VirtualChar CurrentChar => Position < Text.Length ? Text[Position] : new VirtualChar((char)0, default);
 
-        public ImmutableArray<VirtualChar> GetSubPatternToCurrentPos(int start)
+        public VirtualCharSequence GetSubPatternToCurrentPos(int start)
             => GetSubPattern(start, Position);
 
-        public ImmutableArray<VirtualChar> GetSubPattern(int start, int end)
-        {
-            var result = ArrayBuilder<VirtualChar>.GetInstance(end - start);
-            for (var i = start; i < end; i++)
-            {
-                result.Add(Text[i]);
-            }
-
-            return result.ToImmutableAndFree();
-        }
```

Comment:
Large source of allocations removed.

Reply:
now, when a Token/Trivia want to point at a range of characters, they can do it in an alloc-free manner.  They just et a VirtualCharSequence (A struct) which is a sub-span of the original VirtualCharSequence.